### PR TITLE
Deprecate --graph flag; Replace with --data-root

### PR DESCRIPTION
**Pull request by jlhawn**
This patch deprecates the `-g` or `--graph` flag on the `dockerd` or `docker daemon` command in favor of the more descriptive `--data-root` flag.

The name "graph" is a legacy term from long ago when there used to be a directory at the default location `/var/lib/docker/graph`. However, the flag would indicate the path of the parent directory of the "graph" directory which contains not only image data but also data for volumes, containers, and networks. In the most recent version of docker, this directory also contains swarm cluster state and node certificates.

Documentation for the `--graph` option has been updated to use the `--data-root` flag instead. A deprecation notice has also been added.

### CODE REVIEW

**Comment by shykes** [comment]
As the guy who first used the term "graph" in this context (because of the git-like graph relationship between layers), I agree that it's confusing and should be deprecated.

As long as this introduces absolutely no interface regression, I am in favor.

File: docs/reference/commandline/dockerd.md
Comment (StefanScherer) [inline docs/reference/commandline/dockerd.md]: comma is missing

File: docs/deprecated.md
Comment (AkihiroSuda) [inline docs/deprecated.md]: hyperlink wrong: v1.11.0->v1.14.0

Comment (jlhawn) [inline docs/deprecated.md]: Is it okay to set it to a non-existing link?

Comment (jlhawn) [inline docs/deprecated.md]: I'll just remove the link for now.

**Comment by AkihiroSuda** [comment]
According to the docs include in this PR, the milestone of this PR seems 1.14.

However, in 1.13, GraphDriver plugin will be moved to out of experimental.
So, we need to decide whether we rename GraphDriver as well or not.
https://github.com/docker/docker/pull/28623

I'm +1 for renaming GraphDriver, but I don't think we should call it "DataRootDriver"...
Perhaps `--storage-driver` and "StorageDriver"?

PTAL @cpuguy83

**Comment by jlhawn** [comment]
> So, we need to decide whether we rename GraphDriver as well or not.

Oh, I've been wanting to do that for about 2 years: https://github.com/docker/docker/pull/9758 but we never got around to doing it. It's also already consistent with the existing `dockerd` flag:

```
  -s, --storage-driver string                 Storage driver to use
```

> I'm +1 for renaming GraphDriver, but I don't think we should call it "DataRootDriver"...
> Perhaps --storage-driver and "StorageDriver"?

I think these are two distinct things: `--data-root` just specifies where the directory is. Your choice of *copy-on-write image/container filesystem storage driver* is independent of that. The same directory will contain subdirectories for `containers`, `image`, `network`, `swarm`, `tmp`, `volumes`, and whatever the name of the storage driver is that you choose.

**Comment by thaJeztah** [comment]
@AkihiroSuda @jlhawn it was attempted twice; https://github.com/docker/docker/pull/22228 and https://github.com/docker/docker/pull/23237, but there were issues on Windows. Maybe they are resolved in the newer Windows versions.

File: contrib/completion/bash/docker
Comment (thaJeztah) [inline contrib/completion/bash/docker]: Hm, would `-d` become confusing with `-D`? Otherwise I suggest to skip adding a shorthand flag

Comment (jlhawn) [inline contrib/completion/bash/docker]: That sounds reasonable. There's also not a shorthand for the `--exec-root` option. Let me know when the maintainers have a consensus on what to do here and I'll update if needed.

Comment (thaJeztah) [inline contrib/completion/bash/docker]: Probably best to remove it initially, it's easier to add later than to deprecate / remove

Comment (jlhawn) [inline contrib/completion/bash/docker]: Okay, removed for now. I'll add it back if needed.

**Comment by estesp** [comment]
SGTM

**Comment by cpuguy83** [comment]
Ok for me, but I don't think we could ever actually remove `--graph`/`-g`

**Review by tianon — APPROVED** [review]
:heart:!

**Comment by thaJeztah** [comment]
@jlhawn can you keep the old flags around as a hidden flag, in addition to the current one?

Also, the `daemon.json` should probably support the old name as well (but we _can_ issue a warning, and deprecated that, so you need to add that to the `deprecated.md`)

**Comment by jlhawn** [comment]
> can you keep the old flags around as a hidden flag, in addition to the current one?
> Also, the daemon.json should probably support the old name as well (but we can issue a warning, and deprecated that, so you need to add that to the deprecated.md)

Is this not already covered by this: https://github.com/docker/docker/pull/28696/files#diff-a348195a1b645d5b477946b1efae728bR173 ? Also the `graph` option in the config file corresponds to the command line option so you should get a warning for that as well right?

**Comment by LK4D4** [comment]
ping @thaJeztah

**Comment by thaJeztah** [comment]
@jlhawn sorry, yes, you're right about the hidden flag. w.r.t. _removing_ the flag, given that this flag has been around since docker < 1.0, we may need to decide to keep it around, and not actually _remove_ it, but open to suggestions there.

There's a think that needs to be addressed in this PR; you forgot to update the name in the `daemon.json` file; https://github.com/docker/docker/blob/d6be0e98027611cfb14a3246ca797bee0936e649/daemon/config.go#L99

To allow both the _old_ and the _new_ name to work; possibly we need to have an _additional_ property, that's mapped to the old (`graph`) name in the daemon.json. If a user uses _that_ option, a deprecation warning should be printed.

In addition; the deprecation warning is currently printed on the commandline. Given that daemon options would usually be set in either a systemd unit file, or through `daemon.json`, I think we should actually _log_ the error, otherwise people will never see the deprecation message

**Comment by jlhawn** [comment]
@thaJeztah I can't quite wrap my head around the code which handles the config file and command line arguments (theres a bunch of stuff which handles resolving conflicts and setting defaults when both are used). Is there an existing example of how to deprecate something from the config file? The only other daemon flag I see which is marked as deprecated is the `--restart` flag which doesn't have a corresponding config file JSON tag.

> I think we should actually log the error, otherwise people will never see the deprecation message

The flag library being used will print the deprecation warning to `stderr`. What is different about manually logging it (besides the formatting)?

**Comment by jlhawn** [comment]
The more I look at it, the more it looks like I might have to "invent" a way of marking and handling deprecated config file options.

**Comment by thaJeztah** [comment]
Yes, I think this is a first for config-file deprecation. It will probably mean something like;

```go
type CommonConfig struct {
	RootDeprecated     string              `json:"graph,omitempty"`
	Root                         string              `json:"data-dir,omitempty"`
}
```

- if the "old" (`graph`) is set; print deprecation warning
- if both "old" and "new" are set; produce error and fail

w.r.t. both "flag" and "config file"; the current behavior is that if both a flag _and_ the same option in the config file are set, that an error is shown.

**Comment by thaJeztah** [comment]
Ouch, I think this conflicted with https://github.com/docker/docker/pull/29673

@jlhawn do you still want to work on this, or would you prefer someone to carry?

**Comment by jlhawn** [comment]
@thaJeztah hopefully I rebased correctly. I did a local build and it seems to be okay. I'm not sure if I did the `config_solaris.go` file correctly though.

File: daemon/config/config_solaris.go
Comment (jlhawn) [inline daemon/config/config_solaris.go]: I removed these because 1) they don't seem to be used anywhere, and 2) files for other operating systems seem to have been moved to the `cmd/dockerd` package while these were left behind. It looks like this should be covered by https://github.com/docker/docker/pull/28696/files#diff-adcf97f8e8b7815a2ec4660c286531ea

File: daemon/config/config.go
Comment (thaJeztah) [inline daemon/config/config.go]: Looks like validation doesn't currently catch specifying both `--graph` and `--data-dir` as flags

**Review by thaJeztah — DISMISSED** [review]
found some minor issues, but otherwise LGTM; let me push the fixes here to get this going again. 

After this is merged, we'll have to check if other documentation needs to be updated https://github.com/docker/docker.github.io/search?p=1&q=graph&type=&utf8=✓

File: daemon/config/config.go
Comment (ehazlett) [inline daemon/config/config.go]: I think adding the flag check (in aeb461ed64758ebbccd0be303ae5a10cdfcc70dd)  is fine without adding more config validation.

Comment (thaJeztah) [inline daemon/config/config.go]: Yes, I added that commit based on this finding :smile: Was writing it up as an example, and decided to save @jlhawn the time and add rebase it, and add it to the PR

**Comment by ehazlett** [comment]
LGTM

**Comment by ehazlett** [comment]
I also agree we should make it hidden and not remove.  The flag has been there a _long_ time :smile:

**Comment by thaJeztah** [comment]
Let me try to get a bit more input, perhaps bring this one up on today's maintainers discussion. Reading back on the `--net` / `--network` flag, we decided to do a "soft deprecate" on that one; https://github.com/docker/docker/pull/23324#issuecomment-224306027, basically just _hiding_ the `--net` flag, but not outputting a deprecation message.

t.b.h., I think _listing_ that as "deprecated" in "deprecated.md" should still be good to do (it makes the change more widely visible). We can add a new section to that document.

**Comment by thaJeztah** [comment]
Also, looking at this; the "`--graph` is deprecated" message is now printed on the command line, not logged; given that daemon options are usually set in an init-script (e.g. the systemd unit file), I'll add it as a warning to the logs as well.

**Review by vdemeester — APPROVED** [review]
LGTM with ` I'll add it as a warning to the logs as well.` from @thaJeztah 👼

**Comment by thaJeztah** [comment]
ok, updated after discussing with @vdemeester, and changed to _hide_ the flag instead of printing a deprecation warning, and only _log_ the warning. I think this is the least disruptive approach, and still discourages using the old flag, but happy to change back if we think otherwise

**Comment by thaJeztah** [comment]
ok, discussed in the maintainers meeting, and there were no objections to this approach, so should be ready to go :smile:

**Comment by vdemeester** [comment]
Ok then, let's be crazy and merge it 👼

**Comment by thaJeztah** [comment]
Thanks @jlhawn !

**Comment by alexellis** [comment]
Although deprecated - is this a breaking change? There are lots of tutorials etc on the internet that advise on using -g to specify a different path/disk for the Docker data area.

**Comment by vdemeester** [comment]
@alexellis the flags are hidden not removed. It won't break anyone.

> These flags were added before Docker 1.0, so will not be _removed_, only
>_hidden_, to discourage their use.

**Pull request merged**
